### PR TITLE
Add Dockerfile filedef

### PR DESCRIPTION
**Pull request by andy5995**
Per @elextr this will require further discussion.

Closes https://github.com/geany/geany/issues/3752

### CODE REVIEW

**Comment by andy5995** [comment]
https://github.com/dockersamples/genai-chatbot/blob/master/docker-bot/Dockerfile

image

**Comment by eht16** [comment]
See also the already filetype configuration for Docker and Docker-Compose files in the wiki: https://wiki.geany.org/config/docker-compose.

It might be the version in the wiki could be updated for recent Docker features but it is already working.

**Comment by andy5995** [comment]
> See also the already filetype configuration for Docker and Docker-Compose files in the wiki: https://wiki.geany.org/config/docker-compose.

You don't want it included with the Geany distribution then?

**Comment by elextr** [comment]
@andy5995 I think the main point of @eht16's reply (putting words in his mouth :-) was that there are significant differences between the Wiki and your filetype, including more keywords in the wiki version, using Python not Sh for syntax etc etc so its unclear which is better and how to combine them.  That would need to be resolved.  I asked for user [comment](https://github.com/geany/geany/issues/3752#issuecomment-1929287021) about how well yours works, so that question is now expanded to which is better, Dockers please comment.

As for putting it in the Geany release, there have always been debates about adding more filetypes, especially for filetypes that no or few Geany devs use, so we cannot support or make decisions about the filetypes adequacy.  Each language makes the download bigger, adds to the filetype menu, and adds to support.  

Personally I have flip/flopped a few times, but given that the "lightweight" bird has flown some time ago I don't think the cost of a small file for adding a custom filetype is material.  Built-in filetypes are another thing.

The menu issue is harder to address, on small screens large menus may go offscreen, and as menus can't be scrolled that means some items can't be selected (there have been a number of issues about that).  I don't have a solution for this as the Geany categories are hard coded (not that they are very good anyway), maybe it should be alphabetical, PRs welcome :grin:.

As for support, thats where the question to the Docker users in general comes in, does it work well enough that you are actually going to use it?

**Comment by eht16** [comment]
Full agree with what @elextr said, these were mostly the words already in my mouth, I just didn't speak it explicitly.

One addition: after this PR we should *not* have a Docker filetype in Geany and the wiki page as it is now. In the end, we should decide on one solution.

I think it might be worth to add the filetype to Geany itself, Dockerfiles are quite common and often used among developers. Then replace the wiki page with a stub page stating that it has been integrated into Geany and add a reference link to the Git version of the configuration file.

As for the support, if we decide to include the filetype in Geany, I can do a proper review. I use Dockerfiles regularly at work and around. Just be patient with me :D.

**Comment by dolik-rce** [comment]
Hi, Docker user here. 

I have tested both filetype definitions, the one from this PR and the one from wiki. As far as I can tell, they are not much different. I have noticed only one visual difference and that is that the configuration from @andy5995 renders strings in `RUN` commands better (because it uses sh lexer). It also has more up-to-date keywords (contains `MAINTAINER`).

The other file from wiki claims to also support docker-compose files. That is IMHO not very good idea, because those formats are completely different and should not be mixed in single configuration file. It can only lead to worse behavior in both filetypes.

As to whether the filetype should be bundled in Geany release: I'd definitely vote to include it. I am using Geany for years (I could probably say decades by now :sweat_smile:) and it never occurred to me to search for Docker filetype support on wiki. If it were installed with geany, I would definitely be a happier user. It is also much easier than having to set this up manually (even though the documentation is great).

So I would vote to merge this (if I had any voting rights :smiley:) and perhaps remove the other version from wiki.

**Comment by elextr** [comment]
@dolik-rce thanks, if only we got more reviews by third parties things would proceed faster.  I don't know how many people consult the wiki to find out configurations etc. Given the number of projects where the _only_ docs are the wiki I would have thought many dedicated open sourcers would consult it, but beginners who installed from the distro (or windows installer) probably won't think to look, YMMV.  And then there are those who forget ... what was I saying?? ;-)

@andy5995 @dolik-rce @eht16 I have some questions to try:

1. would it be better to make `extension=` the canonical docker extension rather than empty, that way new docker files will get that extension automatically?

Given that the lexer is `Sh` and seems to work well

2. why is the `[styling=C]` not [styling=Sh]` ?
3. why is the `tag_parser=C` not `tag_parser=Sh`?

**Comment by andy5995** [comment]
> @dolik-rce thanks, if only we got more reviews by third parties things would proceed faster. I don't know how many people consult the wiki to find out configurations etc. Given the number of projects where the _only_ docs are the wiki I would have thought many dedicated open sourcers would consult it, but beginners who installed from the distro (or windows installer) probably won't think to look, YMMV. And then there are those who forget ... what was I saying?? ;-)
> 
> @andy5995 @dolik-rce @eht16 I have some questions to try:
> 
>     1. would it be better to make `extension=` the canonical docker extension rather than empty, that way new docker files will get that extension automatically?
> 
> 
> Given that the lexer is `Sh` and seems to work well
> 
>     2. why is the `[styling=C]` not [styling=Sh]` ?
> 
>     3. why is the `tag_parser=C` not `tag_parser=Sh`?

Due to lack of understanding and QA on the part of @andy5995 

Thanks for the explanation above, @elextr and for the feedback and testing @dolik-rce

File: data/filedefs/filetypes.Dockerfile.conf [hunk 1]
```diff
@@ -0,0 +1,14 @@
+[styling=Sh]
+
+[keywords]
+primary=ADD ARG CMD COPY ENTRYPOINT ENV EXPOSE FROM HEALTHCHECK LABEL MAINTAINER ONBUILD RUN SHELL STOPSIGNAL USER VOLUME WORKDIR
+
+[lexer_properties=C]
```
Comment (eht16) [inline data/filedefs/filetypes.Dockerfile.conf:6]:
```suggestion
[lexer_properties=Sh]
```
Or am I missing something?
It won't make much difference as the lexer properties for the C lexer will be probably ignored by the Sh lexer. But to be less confusing, I think it inherit from "Sh".

**Comment by eht16** [comment]
Nice.

I think we could add this new filetype and then update the wiki page to have it only a seperate docker-compose filetype and a reference to the GIT version of this Docker filetype. I'm happy to do this after merge.

Regarding the `extension=` setting: what do you think about using lower case? So that new files will be named "untitled.dockerfile". I don't know if there is a correct answer at all but IIRC what I've seen when there are mutliple dockerfiles is rather something like "Dockerfile.dev" or "Dockerfile.production" but we don't support this ordering in Geany.
So, when using extensions, they are usually written in lowercase in my experience.

For the reasons outlined above, I would add to the filetype extension list the pattern "Dockerfile.*" as this is more common than the others in my experience.

Do we need to adjust anything for the Meson build system?

**Comment by andy5995** [comment]
> Do we need to adjust anything for the Meson build system?

No, this is in data/meson.build:

`install_subdir('filedefs',     install_dir: cdata.get('GEANY_DATA_DIR'), exclude_files: 'filetypes.python.in')`

I think this is done now.

**Comment by eht16** [comment]
If there are no objections, I would merge this next weekend and update the Wiki accordingly.

**Comment by eht16** [comment]
One last remark: could we add `AS` to the keywords? It's not strictly a directive on its own but is part of the `FROM` directive, e.g. `FROM debian:bullseye-slim AS builder` and without the highlighting looks a bit broken.

**Comment by andy5995** [comment]
Yes, I'll do that soon.

**Comment by andy5995** [comment]
> Yes, I'll do that soon.

@eht16 I tried and tested but the AS doesn't highlight.

**Comment by eht16** [comment]
Alright, thanks for the attempt. It seems the Scintilla Bash lexer highlights keywords only at the start of a line which is basically fine for real shell scripts.

I think then it is fine as it is.

**Comment by elextr** [comment]
> but the AS doesn't highlight.

Remember its a `sh` lexer, not a Docker lexer.  AFAICT its somewhat picky about context for keywords, beginning of command is good, after that it varies.

**Comment by eht16** [comment]
Thanks!

I updated the wiki page https://wiki.geany.org/config/docker-compose to only provide a custom filetype for Docker-Compose files.

**Pull request merged**